rebound/examples/PureSystemF.hs

```diff
@@ -0,0 +1,280 @@
+-- | An implementation of System F as a (quasi) Pure Type System.
+module PureSystemF where
+
+import Control.Monad (unless)
+import Control.Monad.Except (Except (..), MonadError (..), runExcept)
+import Data.Fin (f0, f1, f2)
+import Data.Vec ((!))
+import Data.Vec qualified as Vec
+import Rebound
+import Rebound.Bind.Local
+import Rebound.MonadNamed qualified as Scoped
+import Rebound.MonadScoped (MonadScopedReader (..), ScopedReader, ScopedReaderT (..), asksS, runScopedReader)
+import Rebound.MonadScoped qualified as Scoped
+import Text.Read (Lexeme (String))
+
+-- | We represent both terms and types using one single
+-- syntactic class. We use one single constructor for variables,
+-- regardless of whether they stand for a term or a
+-- variable. We also use an additional constructor, 'Kind',
+-- which is used to represent the type of types.
+data Exp (n :: Nat) where
+  Var :: Fin n -> Exp n
+  Kind :: Exp n
+  -- Types
+  TAll :: Bind Ty Ty n -> Ty n
+  TArr :: Ty n -> Ty n -> Ty n
+  -- Terms
+  Abs :: Ty n -> Bind Exp Exp n -> Exp n
+  App :: Exp n -> Exp n -> Exp n
+  TAbs :: Bind Ty Exp n -> Exp n
+  TApp :: Exp n -> Ty n -> Exp n
```

Comment:
We still have different constructs for term-/type- level abstractions/applications.